pkg/tfbridge/validate_input_types.go

```diff
@@ -35,215 +35,249 @@ func NewInputValidator(urn resource.URN, schema pschema.PackageSpec) *PulumiInpu
-			if complexSpec != nil {
-				complexProps, ok := complexSpec.Properties[string(objectKey)]
-				if !ok {
-					failures = append(failures, TypeFailure{
-						ResourcePath: propertyPath.String(),
-						Reason: fmt.Sprintf(
-							"property %s is not defined in the schema",
-							string(objectKey),
-						),
-					})
-					continue
-				}
-				objectTypeSpec = &complexProps.TypeSpec
+	switch typeSpec.Type {
+	case "boolean":
+		// The bridge permits coalescing strings to booleans, hence skip strings.
+		if !propertyValue.IsBool() && !propertyValue.IsString() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected a boolean, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "integer", "number":
+		// The bridge permits coalescing strings to numbers, hence skip strings.
+		if !propertyValue.IsNumber() && !propertyValue.IsString() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected a number, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "string":
+		// The bridge permits coalescing numbers and booleans to strings, hence skip these.
+		if !propertyValue.IsString() && !propertyValue.IsNumber() && !propertyValue.IsBool() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected a string, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "array":
+		if propertyValue.IsArray() {
+			if typeSpec.Items == nil {
+				// Unknown item type so nothing more to check.
+				return nil
 			}
-			if objectTypeSpec != nil {
-				if len(objectTypeSpec.OneOf) > 0 {
-					if failure := v.validateOneOf(string(objectKey), objectValue, objectTypeSpec.OneOf, propertyPath); failure != nil {
-						failures = append(failures, failure...)
-					}
-					continue
-				}
-				if failure := v.validateTypeSpec(
-					string(objectKey),
-					objectValue,
-					*objectTypeSpec,
-					propertyPath,
-					requiredObjectProps,
-				); failure != nil {
+			// Check every item against the array element type.
+			failures := []TypeFailure{}
+			for idx, arrayValue := range propertyValue.ArrayValue() {
+				pb := append(propertyPath, idx)
+				failure := v.validatePropertyValue(arrayValue, *typeSpec.Items, pb)
+				if failure != nil {
 					failures = append(failures, failure...)
 				}
 			}
+			return failures
 		}
+		return []TypeFailure{{
+			ResourcePath: propertyPath.String(),
+			Reason: fmt.Sprintf(
+				"expected an array, got %s type",
+				propertyValue.TypeString(),
+			),
+		}}
+	case "object":
+		// This is not really an object but a map type with some element type, which is assumed to be string if
+		// unspecified. Check accordingly. This should be very similar to the "array" case.
+		//
+		// elementTypeSpec := typeSpec.AdditionalProperties
```

Comment:
This is the only case that needs to handle AdditionalProperties; This needs to be filled in I'm just leaving a TODO for lack of time. IT should be very straightforward along the array lines.